src/compressed_tensors/utils/offload.py

```diff
@@ -49,68 +78,166 @@ def get_execution_device(module: Module) -> torch.device:
+
+
+def update_offload_data(
+    module: torch.nn.Module,
+    name: str,
+    data: Optional[torch.Tensor],
+):
+    """
+    Update the data of an existing parameter and its offload dict. Supports both
+    parameters of offloaded modules and non-offloaded modules
+
+    :param module: module containing the parameter to update
+    :param name: name of module parameter to update
+    :param data: tensor to update parameter with
+    """
+    param = getattr(module, name)
 
-    dtype = parameter.dtype
-    parameter.data = new_param_data.to(device).to(dtype)
+    # copy data into onloaded parameter if applicable
+    if param.device != "meta":
+        param.data.copy_(data)
 
-    if offloaded:
-        prefix_dict = module._hf_hook.weights_map.dataset
-        prefix = module._hf_hook.weights_map.prefix
-        prefix_dict[f"{prefix}{param_name}"] = new_param_data.to(offload_device).to(
-            dtype
-        )
+    # update offload dict
+    if has_offloaded_params(module):
+        weights_map = module._hf_hook.weights_map
+
+        # for upstreaming, better to add write capabilities to weight map classes first
+        if isinstance(weights_map, PrefixedDataset):
+            dataset = getattr_chain(module, "module._hf_hook.weights_map.dataset", None)
+            if dataset is not None:
+                prefix = module._hf_hook.weights_map.prefix
+                key = f"{prefix}{name}"
+
+                offload_device = (
+                    dataset[key].device
+                    if key in dataset
+                    else next(dataset.values()).device
+                )
+                dataset[key] = param.data.to(device=offload_device)
+
+        if isinstance(weights_map, OffloadedWeightsLoader):
+            raise NotImplementedError()
+
+        else:
+            raise NotImplementedError()
+
+
+def delete_offload_parameter(module: torch.nn.Module, name: str):
+    """
+    Delete a module from a module which may be offloaded
+
+    :param module: maybe offloaded module
+    :param name: name of parameter being deleted
+    """
+    delattr(module, name)
+
+    if has_offloaded_params(module):
+        weights_map = module._hf_hook.weights_map
+
+        # for upstreaming, better to add write capabilities to weight map classes first
+        if isinstance(weights_map, PrefixedDataset):
+            dataset = weights_map.dataset
+            prefix = weights_map.prefix
+            if dataset is not None:
+                del dataset[f"{prefix}{name}"]
+
+        elif isinstance(weights_map, OffloadedWeightsLoader):
+            raise NotImplementedError()
+
+        elif weights_map is not None:
+            raise NotImplementedError(
+                f"Cannot delete parameter from weights_map of type {type(weights_map)}"
+            )
+
+
+""" Upstreamed Functions """
+
+
+# introduced in accelerate v1.1.0
+@check_accelerate(fallback=False)
+def has_offloaded_params(module: torch.nn.Module) -> bool:
+    """
```

Comment:
is the docstring style different because this function has been upstreamed?

Reply:
Yes, maybe one day in the future we'll depreciate the mirror function and point people to using the upstreamed function